### PR TITLE
Improve documentation on passing objects across RPC

### PR DESCRIPTION
**Pull request by martin-fleck-at**
#### What it does

Improves documentation.
Fixes https://github.com/eclipse-theia/theia/issues/13224

#### How to test

- Read and see if it is better than before ;-)

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: doc/Plugin-API.md
Comment (jcortell68) [inline doc/Plugin-API.md:156]: see note above regarding `class`

File: doc/Plugin-API.md
Comment (jcortell68) [inline doc/Plugin-API.md:169]: This sentence needs some massaging, IMO. 

As part of that, I would say that the API involves transferring data back and forth that is housed in non-serializable objects, due to the presence of functions in those objects and/or references to other such objects. The objects need to be "dumbed down" to DOT types.

File: doc/Plugin-API.md
Comment (jcortell68) [inline doc/Plugin-API.md:170]: I guess this is the part that has me a bit baffled (still), and why I didn't feel comfortable drafting these improvements. Above we mention the use of MsgPackExtensionManager for marshalling custom non-DTO objects, but here we point out this other technique of _explicitly_ defining DTO counterpart types and converting between them. As a reader, I'm wondering...why? Is it historical (predates MsgPackExtensionManager)? Is there some benefit to doing it explicitly this way in some cases? Is it _necessary_ to do it this way in some cases? Which approach should I take?

File: doc/Plugin-API.md
Comment (martin-fleck-at) [reply doc/Plugin-API.md:75]: Yeah, good point. They truly count as JavaScript built-in objects and are handled by [msgpackr natively](https://github.com/kriszyp/msgpackr/blob/master/pack.js#L419) as well. I'll changed the sentence as you suggested.

File: doc/Plugin-API.md
Comment (martin-fleck-at) [reply doc/Plugin-API.md:76]: I used the word class here because that is what is used as key by msgpackr when adding an extension, see https://github.com/kriszyp/msgpackr?tab=readme-ov-file#custom-extensions. We probably could also use the word `type` here as what you really need to provide is a constructor (function) so that we can make an `instanceof` check (see https://github.com/kriszyp/msgpackr/blob/master/pack.js#L444).

With regard to your comment below, I'll try to discuss with other people what the best option is. I don't want to change it to "non-DTO object" if we actually would suggest to "replace" the usage of DTO objects with a custom serialization approach.

File: doc/Plugin-API.md
Comment (martin-fleck-at) [reply doc/Plugin-API.md:170]: You are right, I believe in theory you could implement all DTOs through the msgpackr extension mechanism and avoid the explicit conversion from domain objects to DTOs and vice versa. I think there are probably several reasons that this is not the case for Theia API:
- msgpackr was only added later and we just added custom serializations for the classes that had that already
- API compliance: VS Code is also using [DTOs](https://github.com/microsoft/vscode/blob/cc7a0865e407e8b5eeaaf0b5cbda6f06f8a99d8b/src/vs/workbench/api/common/extHost.protocol.ts) so it is easier to align with their codebase if we keep the same/a similar system
- msgpackr only allows to register extensions globally so they leak into all connections whether they are related to the custom API or not, see also https://github.com/kriszyp/msgpackr/issues/93. So every single message that is sent may need to be checked against all registered extensions which may impact performance.
- msgpackr only supports [up to 100 extensions](https://github.com/kriszyp/msgpackr?tab=readme-ov-file#custom-extensions) (may not be a big problem for most APIs).

So I think it only makes sense to use msgpackr extensions for very common classes but use DTOs for the actual API. @tsmaeder What do you think about this?

File: doc/Plugin-API.md
Comment (jcortell68) [reply doc/Plugin-API.md:76]: Ah, ok. Makes sense to me.

File: doc/Plugin-API.md
Comment (jcortell68) [reply doc/Plugin-API.md:170]: It seems good to me...as long as it's in the markdown file and not just this PR comment :-)

I hadn't commented on your response since it looked like you were looking for confirmation from @tsmaeder . If you got that confirmation somehow, then yes...please include that insight in the markdown.

Comment (tsmaeder) [reply doc/Plugin-API.md:170]: Sorry, I kind of missed that one. The main problem with this, IMO, is that we don't know the context in which an object is used in a msgpackr extension. We would have to detect the correct type of object purely from the shape of the object. With 10s or 100s of different object, this seems difficult to maintain correctly whereas if we do the conversion manually, we _know_ the type of object we need to convert and the conversion code is close to its usage. Also, we get instances of interfaces from user code sometimes which we need to send to the front end: these objects may contain extra fields which fool our proposed type detection algorithm into thinking it's the wrong type of object. Or a user might implement an interface with a private field (`_myPrivateField`) and a getter (`myPublicGetter`). You can't detect this from shape of the object if you don't know the type.

Comment (martin-fleck-at) [reply doc/Plugin-API.md]: Thank you for your take on that @tsmaeder! I pushed an update that now should better explain when we use the msgpackr extensions and when we use custom DTOs.

**Comment by martin-fleck-at** [comment]
@tsmaeder @jcortell68 I pushed an updated version of the documentation. It would be really great if you could have a very quick read through it to see if there is still something unclear or missing from your point of view. Thank you!

**Comment by jcortell68** [comment]
> @tsmaeder @jcortell68 I pushed an updated version of the documentation. It would be really great if you could have a very quick read through it to see if there is still something unclear or missing from your point of view. Thank you!

The new text/description looks perfect to me. Thank you! I know this was a bit of a pain to pin down, but if it was difficult for us to do that, imagine the Theia newbie working on API trying to figure that out on his own. This will help a lot.

**Review by jcortell68 — APPROVED** [review]

**Comment by jcortell68** [comment]
> @jcortell68 @tsmaeder If the documentation looks good to you, please approve the PR so I can merge it, thank you!

Ah. Didn't think my approval was technically needed since I'm not a maintainer/comitter in the project. Maybe Theia works differently than other open source projects I've worked with. Done.

**Comment by martin-fleck-at** [comment]
@jcortell68 It is technically not needed but I like to get a clean state on the PR before merging it, also for history-preserving reasons ;-)

**Review by jfaltermeier — APPROVED** [review]

**Pull request merged**
